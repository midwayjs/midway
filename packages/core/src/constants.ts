/**
 * 静态参数
 *
 */
export const KEYS = {
  OBJECTS_ELEMENT: 'objects',
  OBJECT_ELEMENT: 'object',
  IMPORT_ELEMENT: 'import',
  PROPERTY_ELEMENT: 'property',
  LIST_ELEMENT: 'list',
  MAP_ELEMENT: 'map',
  ENTRY_ELEMENT: 'entry',
  VALUE_ELEMENT: 'value',
  PROPS_ELEMENT: 'props',
  PROP_ELEMENT: 'prop',
  SET_ELEMENT: 'set',
  CONSTRUCTOR_ARG_ELEMENT: 'constructor-arg',
  REF_ELEMENT: 'ref',
  JSON_ELEMENT: 'json',
  CONFIGURATION_ELEMENT: 'configuration',

  ID_ATTRIBUTE: 'id',
  PATH_ATTRIBUTE: 'path',
  DIRECT_ATTRIBUTE: 'direct',
  AUTOWIRE_ATTRIBUTE: 'autowire',
  ASYNC_ATTRIBUTE: 'async',
  NAME_ATTRIBUTE: 'name',
  REF_ATTRIBUTE: 'ref',
  KEY_ATTRIBUTE: 'key',
  VALUE_ATTRIBUTE: 'value',
  TYPE_ATTRIBUTE: 'type',
  EXTERNAL_ATTRIBUTE: 'external',
  OBJECT_ATTRIBUTE: 'object',
  RESOURCE_ATTRIBUTE: 'resource',
  SCOPE_ATTRIBUTE: 'scope',

  ASPECT_ELEMENT: 'aspect',
  AROUND_ELEMENT: 'around',
  EXPRESSION_ATTRIBUTE: 'expression',
  EXECUTE_ATTRIBUTE: 'execute',
};

export const FUNCTION_INJECT_KEY = 'midway:function_inject_key';
export const MIDWAY_LOGGER_WRITEABLE_DIR = 'MIDWAY_LOGGER_WRITEABLE_DIR';
export const REQUEST_CTX_KEY = 'ctx';
export const REQUEST_OBJ_CTX_KEY = '_req_ctx';
export const HTTP_SERVER_KEY = '_midway_http_server';
export const REQUEST_CTX_LOGGER_CACHE_KEY = '_midway_ctx_logger_cache';

export const ASYNC_CONTEXT_KEY = Symbol('ASYNC_CONTEXT_KEY');
export const ASYNC_CONTEXT_MANAGER_KEY = 'MIDWAY_ASYNC_CONTEXT_MANAGER_KEY';

export const DEFAULT_PATTERN = [
  '**/**.ts',
  '**/**.js',
  '**/**.mts',
  '**/**.mjs',
  '**/**.cts',
  '**/**.cjs',
];

export const IGNORE_PATTERN = ['**/**.d.ts', '**/**.d.mts', '**/**.d.cts'];
