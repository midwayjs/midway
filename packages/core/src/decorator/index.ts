// common
export * from './common/provide';
export * from './common/inject';
export * from './common/pipeline';
export * from './common/aspect';
export * from './common/autoload';
export * from './common/configuration';
export * from './common/objectDef';
export * from './common/framework';
export * from './common/filter';
export * from './common/middleware';
export * from './common/guard';
export * from './common/pipe';
export * from './common/mock';

// faas
export * from './faas/serverlessTrigger';

// web
export * from './web/controller';
export * from './web/paramMapping';
export * from './web/requestMapping';
export * from './web/response';

// other
export * from './constant';
export * from './decoratorManager';

// microservice
export * from './microservice/consumer';
export * from './microservice/provider';
export * from './microservice/rabbitmqListener';
export * from './microservice/kafkaListener';

// rpc
export * from './rpc/hsf';

// task
export * from './task/queue';
export * from './task/task';
export * from './task/taskLocal';
export * from './task/schedule';

// ws
export * from './ws/webSocketController';
export * from './ws/webSocketEvent';
