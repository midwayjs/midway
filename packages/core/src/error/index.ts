export * from './base';
export { HttpStatus, httpError } from './http';
export * from './framework';
