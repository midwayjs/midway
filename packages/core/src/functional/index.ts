export { defineConfiguration } from './configuration';
export * from './hooks';
export * from './interface';
