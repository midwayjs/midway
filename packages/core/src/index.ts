export {
  ObjectIdentifier,
  ObjectDefinitionOptions,
  IManagedInstance,
  ScopeEnum,
  saveClassMetadata,
  attachClassMetadata,
  getClassMetadata,
  saveMethodDataToClass,
  attachMethodDataToClass,
  getMethodDataFromClass,
  listMethodDataFromClass,
  saveMethodMetadata,
  attachMethodMetadata,
  getMethodMetadata,
  savePropertyDataToClass,
  attachPropertyDataToClass,
  getPropertyDataFromClass,
  listPropertyDataFromClass,
  savePropertyMetadata,
  attachPropertyMetadata,
  getPropertyMetadata,
  savePreloadModule,
  listPreloadModule,
  saveModule,
  listModule,
  resetModule,
  clearAllModule,
  getParamNames,
  getProviderId,
  getObjectDefinition,
  classNamed,
  generateProvideId
} from '@midwayjs/decorator';
export * from './interface';
export { ContainerLoader } from './loader';
export { MidwayContainer } from './context/midwayContainer';
export { MidwayRequestContainer } from './context/requestContainer';
export { BaseFramework } from './baseFramework';
export * from './context/providerWrapper';
export * from './common/constants';
export {
  safelyGet,
  safeRequire,
  parsePrefix,
} from './common/util';
export * from './features';
export * from './util/webRouterParam';
export { plainToClass, classToPlain } from 'class-transformer';
