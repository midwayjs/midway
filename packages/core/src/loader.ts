import * as path from 'path';
import { MidwayContainer } from './context/midwayContainer';
import {
  CONFIGURATION_KEY,
  getProviderId,
  listModule,
} from '@midwayjs/decorator';
import { ILifeCycle, IMidwayContainer } from './interface';

function buildLoadDir(baseDir, dir) {
  if (!path.isAbsolute(dir)) {
    return path.join(baseDir, dir);
  }
  return dir;
}

/**
 * @deprecated
 */
export class ContainerLoader {
  baseDir;
  applicationContext: MidwayContainer;
  isTsMode;
  preloadModules;
  disableConflictCheck: boolean;

  constructor({
    baseDir,
    isTsMode = true,
    preloadModules = [],
    disableConflictCheck = true,
  }) {
    this.baseDir = baseDir;
    this.isTsMode = isTsMode;
    this.preloadModules = preloadModules;
    this.disableConflictCheck = disableConflictCheck;
  }

  initialize() {
    this.applicationContext = new MidwayContainer(this.baseDir, undefined);
    this.applicationContext.disableConflictCheck = this.disableConflictCheck;
    this.applicationContext.registerObject('baseDir', this.baseDir);
    this.applicationContext.registerObject('isTsMode', this.isTsMode);
  }

  getApplicationContext(): IMidwayContainer {
    return this.applicationContext;
  }

  registerHook(hookKey, hookHandler) {
    this.applicationContext.registerDataHandler(hookKey, hookHandler);
  }

  loadDirectory(
    loadOpts: {
      baseDir?: string;
      loadDir?: string[];
      disableAutoLoad?: boolean;
      pattern?: string | string[];
      ignore?: string | string[];
    } = {}
  ) {
    if (!this.isTsMode && loadOpts.disableAutoLoad === undefined) {
      // disable auto load in js mode by default
      loadOpts.disableAutoLoad = true;
    }

    if (loadOpts.disableAutoLoad) return;

    // use baseDir in parameter first
    const baseDir = loadOpts.baseDir || this.baseDir;
    const defaultLoadDir = this.isTsMode ? [baseDir] : [];
    this.applicationContext.load({
      loadDir: (loadOpts.loadDir || defaultLoadDir).map(dir => {
        return buildLoadDir(baseDir, dir);
      }),
      pattern: loadOpts.pattern,
      ignore: loadOpts.ignore,
    });

    if (this.preloadModules && this.preloadModules.length) {
      for (const preloadModule of this.preloadModules) {
        this.applicationContext.bindClass(preloadModule);
      }
    }
  }

  async refresh() {
    await this.applicationContext.ready();
    await this.loadLifeCycles();
  }

  async stop() {
    await this.applicationContext.stop();
  }

  async loadLifeCycles() {
    // agent 不加载生命周期
    const cycles: Array<{ target: any; namespace: string }> = listModule(
      CONFIGURATION_KEY
    );
    for (const cycle of cycles) {
      const providerId = getProviderId(cycle.target);
      const inst = await this.getApplicationContext().getAsync<ILifeCycle>(
        providerId
      );
      if (typeof inst.onReady === 'function') {
        /**
         * 让组件能正确获取到 bind 之后 registerObject 的对象有三个方法
         * 1、在 load 之后修改 bind，不太可行
         * 2、每次 getAsync 的时候，去掉 namespace，同时还要查找当前全局的变量，性能差
         * 3、一般只会在 onReady 的地方执行 registerObject（否则没有全局的意义），这个取巧的办法就是 onReady 传入一个代理，其中绑定当前的 namespace
         */
        await inst.onReady(
          new Proxy(this.getApplicationContext(), {
            get: function (target, prop, receiver) {
              if (prop === 'getCurrentNamespace' && cycle.namespace) {
                return () => {
                  return cycle.namespace;
                };
              }
              return Reflect.get(target, prop, receiver);
            },
          })
        );
      }
    }
  }
}
