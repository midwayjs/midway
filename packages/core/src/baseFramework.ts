import {
  IConfigurationOptions,
  IMidwayApplication,
  IMidwayBootstrapOptions,
  IMidwayContainer,
  IMidwayFramework,
  MidwayFrameworkType,
  MidwayProcessTypeEnum,
} from './interface';
import { ContainerLoader } from './';
import { APPLICATION_KEY } from '@midwayjs/decorator';

export abstract class BaseFramework<
  APP extends IMidwayApplication,
  T extends IConfigurationOptions
> implements IMidwayFramework<APP, T> {
  protected isTsMode = true;
  protected baseDir: string;
  protected appDir: string;
  protected containerLoader: ContainerLoader;
  public configurationOptions: T;
  public app: APP;

  public configure(options: T): BaseFramework<APP, T> {
    this.configurationOptions = options;
    return this;
  }

  public async initialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {
    this.baseDir = options.baseDir;
    this.appDir = options.appDir;

    this.containerLoader = new ContainerLoader({
      baseDir: this.baseDir,
      isTsMode: this.isTsMode,
      preloadModules: options.preloadModules || [],
    });

    /**
     * initialize containerLoader and initialize ioc container instance
     */
    await this.beforeInitialize(options);
    this.containerLoader.initialize();

    /**
     * load directory and bind files to ioc container
     */
    await this.beforeDirectoryLoad(options);
    const applicationContext = this.containerLoader.getApplicationContext();
    applicationContext.registerObject('baseDir', this.baseDir);
    applicationContext.registerObject('appDir', this.appDir);
    // 如果没有关闭autoLoad 则进行load
    this.containerLoader.loadDirectory(options);

    // register app
    this.containerLoader.registerHook(APPLICATION_KEY, () => {
      return this.getApplication();
    });

    await this.afterDirectoryLoad(options);

    if (!this.app.getApplicationContext) {
      this.app = this.defineApplicationProperties(this.app);
    }

    /**
     * start to load configuration and lifeCycle
     */
    await this.containerLoader.refresh();
    await this.afterInitialize(options);
  }

  public getApplicationContext(): IMidwayContainer {
    return this.containerLoader.getApplicationContext();
  }

  public getConfiguration(key?: string): any {
    return this.getApplicationContext()
      .getConfigService()
      .getConfiguration(key);
  }

  public getCurrentEnvironment() {
    return this.getApplicationContext()
      .getEnvironmentService()
      .getCurrentEnvironment();
  }

  public abstract getFrameworkType(): MidwayFrameworkType;

  public abstract getApplication(): APP;

  public abstract run(): Promise<void>;

  public async stop(): Promise<void> {
    await this.beforeStop();
    await this.containerLoader.stop();
  }

  protected defineApplicationProperties(app: APP): APP {
    return Object.assign(app, {
      getBaseDir: () => {
        return this.baseDir;
      },

      getAppDir: () => {
        return this.appDir;
      },

      getEnv: () => {
        return this.getApplicationContext()
          .getEnvironmentService()
          .getCurrentEnvironment();
      },

      getConfig: (key?: string) => {
        return this.getApplicationContext()
          .getConfigService()
          .getConfiguration(key);
      },

      getFrameworkType: () => {
        return this.getFrameworkType();
      },

      getProcessType: () => {
        return MidwayProcessTypeEnum.APPLICATION;
      },
    });
  }

  protected async beforeStop(): Promise<void> {}

  protected async beforeInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async beforeDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async afterDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected abstract async afterInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void>;
}
