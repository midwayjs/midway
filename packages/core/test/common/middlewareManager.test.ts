import { ContextMiddlewareManager } from '../../src';

const mw1 = async (ctx, next) => {};
mw1._name = 'mw1';

const mw2 = async (ctx, next) => {};
mw2._name = 'mw2';

const mw3 = class MW3 {resolve() {}} as any;
const mw4 = class MW4 {resolve() {}} as any;

function mw5(){}
function mw6(){}

describe('test/common/middlewareManager.test.ts', function () {
  it('should test manager like array insert first', function () {
    const manager = new ContextMiddlewareManager();
    // insert first
    manager.insertFirst(mw1);
    manager.insertFirst(mw2);
    manager.insertFirst([mw4, mw3]);

    expect(manager.length).toEqual(4);
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test manager like array insert last', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertLast([mw2, mw3]);
    manager.insertLast(mw4);
    manager.insertLast(mw5);
    manager.insertFirst(mw1);

    expect(manager.length).toEqual(5);
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test manager like array insert before', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst(mw1);
    manager.insertBefore([mw2, mw3], mw1);
    manager.insertBefore(mw5, 2);
    manager.insertBefore(mw6, 'MW3');

    expect(manager.length).toEqual(5);
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test manager like array insert after', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst(mw1);
    manager.insertAfter([mw2, mw3], mw1);
    manager.insertBefore(mw5, 2);
    manager.insertAfter(mw6, 'mw1');

    expect(manager.length).toEqual(5);
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test idx greater then the length of array', function () {
    const manager = new ContextMiddlewareManager();
    manager.insertAfter(mw1, 5);
    expect(manager.length).toEqual(1);

    manager.insertBefore(mw2, -1);
    manager.insertBefore(mw3, -2);
    expect(manager.length).toEqual(3);

    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test find middleware', function () {
    const manager = new ContextMiddlewareManager();
    manager.push(mw1);
    manager.push(mw2);
    manager.insertLast([mw4, mw3]);

    expect(manager.findItemIndex('MW4')).toEqual(2);
    expect(manager.findItemIndex(mw4)).toEqual(2);
    expect(manager.findItemIndex(2)).toEqual(2);
    expect(manager.findItemIndex(mw5)).toEqual(-1);

    expect(manager.findItem(mw3)).toEqual(mw3);
    expect(manager.findItem('MW3')).toEqual(mw3);
    expect(manager.findItem(-1)).toBeUndefined();
    expect(manager.findItem(10)).toBeUndefined();
  });

  it('should get name from middleware', function () {
    const manager = new ContextMiddlewareManager();

    expect(manager.getMiddlewareName(mw1)).toEqual('mw1');
    expect(manager.getMiddlewareName(mw3)).toEqual('MW3');
    expect(manager.getMiddlewareName(mw5)).toEqual('mw5');
  });

  it('should remove middleware', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst(mw1);
    manager.insertFirst(mw2);
    manager.insertFirst([mw4, mw3]);

    manager.remove(mw1);
    expect(manager.length).toEqual(3);
    expect(manager.getNames()).toMatchSnapshot();

    manager.remove('MW3');
    expect(manager.length).toEqual(2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.remove(1);
    expect(manager.length).toEqual(1);
    expect(manager.getNames()).toMatchSnapshot();

    manager.remove(5);
    expect(manager.length).toEqual(1);
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test find and insert after', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst([mw1, mw2, mw3, mw4, mw5, mw6]);

    manager.findAndInsertAfter(mw3, mw5);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter(mw3, 'mw2');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter('mw2', 'MW4');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter('MW4', 4);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter(mw3, 'mw7');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter('mw8', mw2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter(mw2, mw2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter('mw2', mw2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertAfter(mw2, 'mw2');
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test find and insert before', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst([mw1, mw2, mw3, mw4, mw5, mw6]);

    manager.findAndInsertBefore(mw3, mw5);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore(mw3, 'mw2');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore('mw6', 'MW4');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore('MW4', 1);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore(mw3, 'mw7');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore('mw8', mw2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore(mw2, mw2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore('mw2', mw2);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertBefore(mw2, 'mw2');
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test find and insert first', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst([mw1, mw2, mw3, mw4, mw5, mw6]);

    manager.findAndInsertFirst(mw5);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertFirst('mw2');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertFirst('mw7');
    expect(manager.getNames()).toMatchSnapshot();
  });

  it('should test find and insert last', function () {
    const manager = new ContextMiddlewareManager();

    manager.insertFirst([mw1, mw2, mw3, mw4, mw5, mw6]);

    manager.findAndInsertLast(mw5);
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertLast('mw2');
    expect(manager.getNames()).toMatchSnapshot();

    manager.findAndInsertLast('mw7');
    expect(manager.getNames()).toMatchSnapshot();
  });
});
