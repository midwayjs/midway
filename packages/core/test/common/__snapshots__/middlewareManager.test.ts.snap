// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/common/middlewareManager.test.ts should remove middleware 1`] = `
ContextMiddlewareManager [
  "MW4",
  "MW3",
  "mw2",
]
`;

exports[`test/common/middlewareManager.test.ts should remove middleware 2`] = `
ContextMiddlewareManager [
  "MW4",
  "mw2",
]
`;

exports[`test/common/middlewareManager.test.ts should remove middleware 3`] = `
ContextMiddlewareManager [
  "MW4",
]
`;

exports[`test/common/middlewareManager.test.ts should remove middleware 4`] = `
ContextMiddlewareManager [
  "MW4",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 1`] = `
ContextMiddlewareManager [
  "mw1",
  "mw2",
  "MW4",
  "mw5",
  "MW3",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 2`] = `
ContextMiddlewareManager [
  "mw1",
  "mw2",
  "MW3",
  "MW4",
  "mw5",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 3`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "MW4",
  "mw2",
  "mw5",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 4`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw5",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 5`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw5",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 6`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw5",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 7`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw5",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 8`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw5",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert after 9`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw5",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 1`] = `
ContextMiddlewareManager [
  "mw1",
  "mw2",
  "MW4",
  "MW3",
  "mw5",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 2`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "MW4",
  "mw5",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 3`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "mw2",
  "mw6",
  "MW4",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 4`] = `
ContextMiddlewareManager [
  "mw1",
  "MW4",
  "MW3",
  "mw2",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 5`] = `
ContextMiddlewareManager [
  "mw1",
  "MW4",
  "MW3",
  "mw2",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 6`] = `
ContextMiddlewareManager [
  "mw1",
  "MW4",
  "MW3",
  "mw2",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 7`] = `
ContextMiddlewareManager [
  "mw1",
  "MW4",
  "MW3",
  "mw2",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 8`] = `
ContextMiddlewareManager [
  "mw1",
  "MW4",
  "MW3",
  "mw2",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert before 9`] = `
ContextMiddlewareManager [
  "mw1",
  "MW4",
  "MW3",
  "mw2",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert first 1`] = `
ContextMiddlewareManager [
  "mw5",
  "mw1",
  "mw2",
  "MW3",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert first 2`] = `
ContextMiddlewareManager [
  "mw2",
  "mw5",
  "mw1",
  "MW3",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert first 3`] = `
ContextMiddlewareManager [
  "mw2",
  "mw5",
  "mw1",
  "MW3",
  "MW4",
  "mw6",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert last 1`] = `
ContextMiddlewareManager [
  "mw1",
  "mw2",
  "MW3",
  "MW4",
  "mw6",
  "mw5",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert last 2`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "MW4",
  "mw6",
  "mw5",
  "mw2",
]
`;

exports[`test/common/middlewareManager.test.ts should test find and insert last 3`] = `
ContextMiddlewareManager [
  "mw1",
  "MW3",
  "MW4",
  "mw6",
  "mw5",
  "mw2",
]
`;

exports[`test/common/middlewareManager.test.ts should test idx greater then the length of array 1`] = `
ContextMiddlewareManager [
  "MW3",
  "mw2",
  "mw1",
]
`;

exports[`test/common/middlewareManager.test.ts should test manager like array insert after 1`] = `
ContextMiddlewareManager [
  "mw1",
  "mw6",
  "mw2",
  "mw5",
  "MW3",
]
`;

exports[`test/common/middlewareManager.test.ts should test manager like array insert before 1`] = `
ContextMiddlewareManager [
  "mw2",
  "mw6",
  "MW3",
  "mw5",
  "mw1",
]
`;

exports[`test/common/middlewareManager.test.ts should test manager like array insert first 1`] = `
ContextMiddlewareManager [
  "MW4",
  "MW3",
  "mw2",
  "mw1",
]
`;

exports[`test/common/middlewareManager.test.ts should test manager like array insert last 1`] = `
ContextMiddlewareManager [
  "mw1",
  "mw2",
  "MW3",
  "MW4",
  "mw5",
]
`;
